### PR TITLE
test: expand cabinet IR and DSP stability coverage

### LINKED ISSUE #184: test: Cover Cabinet Sim IR loading (84%), convolution_engine.cpp (74%), and wav_loader.cpp (36%)
## Coverage Baseline

From the PR #178 coverage report:

| File | Lines | Functions |
|---|---|---|
| `cabinet_sim.cpp` | 84.6% (88/104) | **83.3%** (10/12) |
| `cabinet_sim.h` | **33.3%** (1/3) | **33.3%** (1/3) |
| `convolution_engine.cpp` | **74.2%** (112/151) | **83.3%** (10/12) |
| `wav_loader.cpp` | **36.5%** (23/63) | **50.0%** (1/2) |

### `wav_loader.cpp` (50% function coverage — 1 uncovered function)
- `load_wav_file(path)` — loading a WAV file from disk for IR cabinet use
- - This is used by `CabinetSim` to load impulse response files
- - - Error paths: file not found, malformed WAV, wrong sample rate, stereo vs. mono
- - - - `resample_to_rate(target_rate)` — resampling loaded IR to match engine sample rate
### `convolution_engine.cpp` (2 uncovered functions)
- `ConvolutionEngine::reset()` — clearing internal state between IR loads
- - `ConvolutionEngine::get_ir_length()` — querying loaded IR length
- - - Latency: processing a block shorter than the IR length
- - - - Overlap-add correctness: verify that splitting input into blocks gives same result as processing all at once
### `cabinet_sim.cpp` (2 uncovered functions)
- `CabinetSim::load_ir_from_file(path)` — the full load path including `WavLoader` + `ConvolutionEngine::load_ir()`
- - `CabinetSim::get_ir_name()` — returns the filename of the currently loaded IR
- - - Mode switch: `set_mode(CONVOLUTION)` vs. `set_mode(PARAMETRIC)`
### `cabinet_sim.h` (33% — only 1/3 methods covered)
- `is_ir_loaded()` — status query
- - `get_ir_duration_ms()` — IR duration in ms
## Tests to Add

In `tests/test_cabinet_sim_ir.cpp` (extend or create):

```cpp
TEST(WavLoader, LoadValidWavFile) {
    // Use a known-good test WAV in tests/fixtures/
    WavLoader loader;
    ASSERT_TRUE(loader.load("/path/to/tests/fixtures/test_ir.wav"));
    ASSERT_GT(loader.get_sample_count(), 0);
    ASSERT_EQ(loader.get_sample_rate(), 48000);
}

TEST(WavLoader, LoadMissingFileReturnsFalse) {
    WavLoader loader;
    ASSERT_FALSE(loader.load("/nonexistent/missing.wav"));
}

TEST(WavLoader, LoadMalformedWavReturnsFalse) {
    WavLoader loader;
    ASSERT_FALSE(loader.load("/path/to/tests/fixtures/not_a_wav.txt"));
}

TEST(WavLoader, ResampleToMatchEngineRate) {
    WavLoader loader;
    loader.load("/path/to/tests/fixtures/test_ir_44khz.wav"); // 44.1kHz source
    auto resampled = loader.resample_to_rate(48000);
    ASSERT_EQ(resampled.sample_rate(), 48000);
    ASSERT_GT(resampled.get_sample_count(), 0);
}

TEST(ConvolutionEngine, ResetClearsInternalState) {
    ConvolutionEngine conv;
    conv.load_ir(test_ir_data);
    auto out1 = conv.process(test_input);
    conv.reset();
    auto out2 = conv.process(test_input);
    ASSERT_NE(out1, out2); // reset should affect output
}

TEST(ConvolutionEngine, OverlapAddBlockConsistency) {
    ConvolutionEngine conv;
    conv.load_ir(test_ir_data);
    // Process all at once
    auto result_whole = conv.process(large_input);
    // Process in blocks
    conv.reset();
    std::vector<float> result_blocks;
    for (int i = 0; i < large_input.size(); i += block_size) {
        auto block_out = conv.process(block_view(large_input, i, block_size));
        result_blocks.insert(result_blocks.end(), block_out.begin(), block_out.end());
    }
    // Results should be identical (within floating-point tolerance)
    for (int i = 0; i < result_whole.size(); i++) {
        ASSERT_NEAR(result_whole[i], result_blocks[i], 1e-5f);
    }
}

TEST(CabinetSim, LoadIRFromFile) {
    CabinetSim cab;
    cab.set_sample_rate(48000);
    ASSERT_TRUE(cab.load_ir_from_file("/path/to/tests/fixtures/test_ir.wav"));
    ASSERT_TRUE(cab.is_ir_loaded());
}

TEST(CabinetSim, GetIRName) {
    CabinetSim cab;
    cab.set_sample_rate(48000);
    cab.load_ir_from_file("/path/to/tests/fixtures/Marshall.wav");
    ASSERT_EQ(cab.get_ir_name(), "Marshall.wav");
}

TEST(CabinetSim, ModeSwitch) {
    CabinetSim cab;
    cab.set_mode(CabinetSim::Mode::CONVOLUTION);
    ASSERT_EQ(cab.get_mode(), CabinetSim::Mode::CONVOLUTION);
    cab.set_mode(CabinetSim::Mode::PARAMETRIC);
    ASSERT_EQ(cab.get_mode(), CabinetSim::Mode::PARAMETRIC);
}

TEST(CabinetSimHeader, IsIRLoaded) {
    CabinetSim cab;
    ASSERT_FALSE(cab.is_ir_loaded()); // initially empty
    cab.load_ir_from_file("/path/to/tests/fixtures/test_ir.wav");
    ASSERT_TRUE(cab.is_ir_loaded());
}

TEST(CabinetSimHeader, GetIRDurationMs) {
    CabinetSim cab;
    cab.set_sample_rate(48000);
    cab.load_ir_from_file("/path/to/tests/fixtures/1sec_ir.wav"); // 1 second
    ASSERT_NEAR(cab.get_ir_duration_ms(), 1000.0f, 1.0f);
}
```

Note: The `test_ir.wav` files should be included in `tests/fixtures/` and be real WAV files (even if synthetic/generated), not placeholder content.

### PR DESCRIPTION
**Pull request by YASHcode-IIITV**
## What does this PR do?

Expands cabinet IR and DSP stability test coverage for the cabinet simulation pipeline.

Adds regression and edge-case tests for:

* Missing IR file handling
* Malformed WAV file rejection
* Long-run numerical stability
* Silence propagation correctness
* Unit impulse identity behavior
* Delayed impulse convolution behavior

## Related Issue

Fixes #184 

## Type of Change

* [ ] 🐛 Bug fix
* [ ] ✨ New feature
* [ ] ♻️ Refactor / code cleanup
* [ ] 📝 Documentation
* [x] ✅ Tests
* [ ] 🔧 Build / CI / Configuration

## How Was This Tested?

* Platform(s) tested on:

  * Windows 11 x64
  * MSVC + CMake + Ninja

* Test command run:
  `build\amplitron-tests.exe`

* Manual test steps (if applicable):

  * Built project in Release mode
  * Executed full test suite after adding new cabinet IR tests
  * Verified all tests pass successfully

Results:

* 232 passed
* 0 failed

## Checklist

* [x] Code compiles and builds successfully on my platform
* [x] All existing tests pass (`./amplitron-tests`)
* [x] New tests added for new functionality (if applicable)
* [ ] Documentation updated (README, code comments) if applicable
* [x] No blocking calls on the audio thread (no `std::mutex::lock()` on the hot path)
* [x] Tested on: Windows 11 x64

## Screenshots / Demo

N/A — test coverage changes only.

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive unit tests for audio subsystems: WAV loading/truncation/resampling, convolution engine behavior, cabinet-simulation IR handling, and effect stability.
  * Expanded coverage for error handling (missing/truncated/malformed files), sample-length limits, state reset/clear and passthrough when disabled, long-run stability, block-size edge cases, and cleanup of temporary test artifacts.

[Review Change Stack](https://app.coderabbit.ai/change-stack/sudip-mondal-2002/Amplitron/pull/207?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by YASHcode-IIITV** [comment]
@sudip-mondal-2002 i have completed the changes and it is ready for the merger

File: tests/test_convolution_engine.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    conv.reset();
+
+    std::vector<float> silent(256, 0.0f);
+
+    conv.process(silent.data(), 256);
+
+    for (float s : silent) {
+        ASSERT_TRUE(std::isfinite(s));
+    }
```
[the remark this answers is not in the document]
Comment (YASHcode-IIITV) [reply tests/test_convolution_engine.cpp:34]: resolved

**Comment by MohitBareja16** [comment]
Screenshot_2026-05-23-17-08-10-67_40deb401b9ffe8e1df2f1cc5ba480b12
Screenshot_2026-05-23-17-07-39-75_40deb401b9ffe8e1df2f1cc5ba480b12
convulation_engine and wav_loader files don't have enough coverage. Make it 90+. Also cabinet_sim is good but try to make it 90+

**Comment by YASHcode-IIITV** [comment]
@MohitBareja16 plz review it .

**Comment by sudip-mondal-2002** [comment]
wav_loader is still red, if possible get it 90+, if not possible let us know why

**Comment by YASHcode-IIITV** [comment]
> wav_loader is still red, if possible get it 90+, if not possible let us know why

image
you can see that the test time is of yesterday at 12 pm can you check the current report with the current time test .

**Review by sudip-mondal-2002 — CHANGES_REQUESTED** [review]
please keep the assets inside `tests/assets` folder

otherwise LGTM

**Review by sudip-mondal-2002 — DISMISSED** [review]
LGTM, but the coverage still looks low on convo eng
image

**Review by sudip-mondal-2002 — APPROVED** [review]
